kodi-config: Hack to work around repo issue following OE-to-LE migration

MilhouseVH opened the pull request:
An OE Kodi 16 installation will, naturally enough, have an Addons*.db that is populated with OE addons linked to the `repository.openelec.tv` repo.

However, following the migration to LE, Kodi 16 adds the same addons to Addons*.db, but now for the `repository.libreelec.tv` repo - the old OE addons are still present in the database, even though the `repository.openelec.tv` repo no longer exists.

So now, for example, there will be two `pvr.demo` addons in the `addon` table, one linked to the OE repo (usually with a low id, ie. 8) and again for the LE repo, but now with a much higher id, eg. 980.

When the user attempts to install the `pvr.demo` addon from the LE repository, Kodi seems to use the first addon it comes across, which is typically the addon with the lowest id and thus the legacy addon linked to the now non-existant `repository.openelec.tv`. Result: nothing happens - no error is shown to the user, or written to the log. The addon simply fails to install, silently.

As a quick fix, this PR simply deletes the OE repository and all associated addons from the latest pre-v21 Addons*.db. It also creates a backup, should the user decide to go back to OE.

Ideally, Kodi 16 would simply ignore those addons linked to repositories that are no longer available/valid.

Merge if nobody finds a better solution to the problem.

stefansaraev commented:
ouch.

lrusak commented:
maybe @stefansaraev has a better idea 😛 

otherwise we gotta do what we gotta do

stefansaraev commented:
I do. delete everything < Addons21.db

oh wait. /me runs

lrusak commented:
We thought of that but add-ons that are disabled will be re-enabled.

stefansaraev commented:
oh. that. yeah.

and btw, you can't fix mysql installs. anyway. right?

I heard v17 will be even more fun. you should not care too much about db migration ;)

lrusak commented:
Yea mysql people are on their own to drop the addons database

Yes, so this is just putting lipstick on a pig

MilhouseVH commented:
I've had to revise the logic slightly, to account for an indecisive user that migrates from OE to LE, returns to OE (automatically adding back the addons/repositories, although now with "higher" the LE ids, meaning any addon problems are now shifted to OE), then maybe later they decide to try LE again... Although LE would probably function correctly as it now has the lower id's it's a bit of a gamble, so with this revised PR the OE addons/repositories will always be removed whenever they are detected.

MilhouseVH commented:
Ping @tamland, if you have any good ideas.

MilhouseVH commented:
Dropped the version check as the queries work with Kodi 17 too.

Unfortunately this hack is only modifying the latest available db which might not be the db Kodi is using, eg. when testing Kodi 16 on a system that previously tested Kodi 17, the Kodi 17 db is always modified not the Kodi 16 db... Is this an acceptable limitation, rather than apply a more invasive fix in Kodi?

Also dropped `repository.unofficial.addon.pro`, as this is a user installed repository that will keep being re-added until it is uninstalled.

MilhouseVH commented:
Updated - now we always fix `Addon20.db`, if it exists, as that's the only database Kodi 16 is interested in. Possibly in an upgrade situation (eg. `Addons19.db`) things could be a little odd on the first boot, but will be fixed on the second and subsequent boots.

For Kodi 17, we'll need to amend the patch to work with whatever Addons database version the supplied Kodi is using. Not ideal, as it means bumping this patch every time Kodi bump Addons db, although fortunately that isn't often. Kodi 17 currently uses `Addons23.db`.

tamland commented:
This is an inherent problem in the database/installer unfortunately. Many things just assume there is one version in one repository. Also, it was never designed to handle a repository just 'disappearing' without being uninstalled. Most of the groundwork needed was already done in jarvis with the new updating functionality, and I'm planning to improve this further. Don't expect any backports though. Far too late for that..

MilhouseVH commented:
Fair enough on the backport. Would be great if something could be done for Kodi 17 though, as repositories appearing/disappearing could become a more common occurrence...
